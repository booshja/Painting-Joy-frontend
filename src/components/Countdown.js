// dependencies
import React, { useEffect, useRef, useState } from "react";
import styled from "styled-components";

const StyledCountdown = styled.p`
    font-family: "News Cycle", sans-serif;
    font-size: 1.2rem;
    font-weight: 700;
    color: #207070;
    letter-spacing: 1px;
    margin-right: 2rem;
    padding-bottom: 1rem;
    align-self: flex-end;
`;

const Countdown = ({ handleCancel }) => {
    // set up ref for setTimeout
    const Ref = useRef(null);
    // set up state
    const [timer, setTimer] = useState("5:00");

    const getTimeRemaining = (e) => {
        // calculate time remaining for countdown clock
        const total = Date.parse(e) - Date.parse(new Date());
        const seconds = Math.floor((total / 1000) % 60);
        const minutes = Math.floor((total / 1000 / 60) % 60);

        return { total, minutes, seconds };
    };

    const startTimer = (e) => {
        // set up timer, start counting
        let { total, minutes, seconds } = getTimeRemaining(e);
        if (total >= 0) {
            // update timer
            setTimer(minutes + ":" + (seconds > 9 ? seconds : "0" + seconds));
        }

        if (total === 0) {
            handleCancel();
        }
    };

    const clearTimer = (e) => {
        // reset timer
        setTimer("5:00");

        if (Ref.current) clearInterval(Ref.current);
        const id = setInterval(() => {
            startTimer(e);
        }, 1000);
        Ref.current = id;
    };

    const getDeadTime = () => {
        // calculate finish time for 5 minutes
        let deadline = new Date();

        deadline.setMinutes(deadline.getMinutes() + 5);
        return deadline;
    };

    useEffect(() => {
        // when component mounts, start timer
        clearTimer(getDeadTime());

        return function cleanup() {
            clearInterval(Ref.current);
        };
    }, []);

    return <StyledCountdown>Time Remaining: {timer}</StyledCountdown>;
};

export default Countdown;
