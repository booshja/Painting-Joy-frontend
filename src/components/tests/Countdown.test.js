import { render } from "@testing-library/react";
import Countdown from "../Countdown";
import "jest-styled-components";

/** Smoke Test */
it("renders without crashing", () => {
    render(<Countdown handleCancel={null} />);
});

/** Snapshot Test */
it("matches snapshot", () => {
    const { asFragment } = render(<Countdown handleCancel={null} />);
    expect(asFragment()).toMatchSnapshot();
});

/** RTL Tests */
it("renders correctly", () => {
    const { getByText } = render(<Countdown handleCancel={null} />);

    expect(getByText("Time Remaining: ", { exact: false })).toBeInTheDocument();
});
