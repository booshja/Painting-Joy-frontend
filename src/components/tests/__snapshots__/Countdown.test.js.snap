// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  font-family: "News Cycle",sans-serif;
  font-size: 1.2rem;
  font-weight: 700;
  color: #207070;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-right: 2rem;
  padding-bottom: 1rem;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
}

<p
    class="c0"
  >
    Time Remaining: 5:00
  </p>
</DocumentFragment>
`;
